src/bin/pg_autoctl/supervisor.c

```diff
@@ -56,22 +74,278 @@ static bool supervisor_restart_service(Supervisor *supervisor,
 									   Service *service,
 									   int status);
 
+static void supervisor_handle_failed_restart(Supervisor *supervisor,
+											 Service *service,
+											 int status);
+
 static bool supervisor_may_restart(Service *service);
 
 static bool supervisor_update_pidfile(Supervisor *supervisor);
 
 
+/*
+ * supervisor_it_* implement a helper to loop through the active services in a
+ * supervisor struct. They can be used either on their own, or via the helper
+ * macro foreach_supervised_service().
+ * Consult the macro for an example of use.
+ */
+static inline Service *
+supervisor_it_get(Supervisor_it *sit)
+{
+	if (sit->index < sit->count)
+	{
+		return &sit->services[sit->index];
+	}
+
+	return NULL;
+}
+
+
+static inline Service *
+supervisor_it_first(Supervisor_it *sit, Supervisor *supervisor)
+{
+	*sit = (Supervisor_it) {
+		.supervisor = supervisor,
+		.services = supervisor->services,
+		.count = supervisor->serviceCount,
+		.index = 0,
+	};
+
+	return supervisor_it_get(sit);
+}
+
+
+static inline Service *
+supervisor_it_next(Supervisor_it *sit)
+{
+	sit->index++;
+
+	/* if we are done with the static services, move to the dynamic */
+	if (sit->index >= sit->count &&
+		sit->services == sit->supervisor->services)
+	{
+		sit->services = sit->supervisor->dynamicServicesEnabled.services;
+		sit->count = sit->supervisor->dynamicServicesEnabled.serviceCount;
+		sit->index = 0;
+	}
+
+	return supervisor_it_get(sit);
+}
+
+
+#define foreach_supervised_service(it, supervisor, service) \
+	for ((service) = supervisor_it_first((it), (supervisor)); \
+		 (service) != NULL; \
+		 (service) = supervisor_it_next(it))
+
+
+/*
+ * supervisor_dynamic_service_enable starts the service and if successfull, adds
+ * it to the dynamicService array.
+ *
+ * Since the services are largely discoverable by name, the name has to be
+ * unique. If it is not, then the service is not started and the function
+ * returns false.
+ * Also since the service runs under supervision, there must be space for it to
+ * be added to the supervisor's dynamic services array. If there is not, then
+ * the service is not started and the function returns false.
+ * The service will not be added if a signal other than "dynamic service" has
+ * been received.
+ *
+ * Returns true if successfully starts the service.
+ */
+bool
+supervisor_dynamic_service_enable(Supervisor *supervisor, Service *service)
+{
+	Service *dynamicService;
+	sigset_t sig_mask;
+	sigset_t sig_mask_orig;
+	int serviceIndex = supervisor->dynamicServicesEnabled.serviceCount;
+	bool started;
+
+	if (supervisor_service_exists(supervisor, service->name))
+	{
+		log_error("Service %s already exists under supervision",
+				  service->name);
```

Comment:
Sure.

What would you like the behaviour to be if it is not know to the OS?

Reply:
In other parts of the code, when we see a stale PID file, we remove the file and then start the service again. Here it's a little more confusing because it's a stale entry in-memory in our array of services. We have to think about race conditions with the disable approach and all in that case I suppose.

Reply:
Yeah, I have thought of those race conditions and cases. A stale PID in this case should indicate a coding error imho.

Otherwise the loop should have removed a failed process. Would you agree with that?

Reply:
Okay now I have to do my homework and follow it through on my side. Getting back to you later ;-)